File: mobile/src/api/answerLocation.ts
import type { LocalAnswer } from '../types/models';

/**
 * The per-answer location provenance both answer endpoints accept, in the
 * backend's wire shape.
 *
 * One module rather than two copies because knowledge-gap answers
 * (POST /questions/{id}/answers) and place-question answers
 * (POST /place-questions/{id}/answers) take the identical optional block --
 * and because the rule that a coordinate is only ever sent as 'gps_live'
 * when there genuinely IS a device fix behind it has to hold for both.
 */
export interface AnswerLocationWire {
  latitude?: number;
  longitude?: number;
  location_source?: string;
  location_accuracy_meters?: number;
  location_captured_at?: string;
  location_label?: string;
  external_place_id?: string;
}

/**
 * Builds the wire block for an answer, or `{}` when the guide never captured
 * a location for it.
 *
 * Omits the fields entirely rather than sending nulls: an absent block means
 * "the client has nothing to say about where this happened", which is exactly
 * what makes the server keep its existing derivation (the knowledge gap's
 * target coordinates, or the place a place question is about). Sending
 * `location_source: 'gps_live'` with no coordinates would be a claim the data
 * cannot back up -- the backend rejects that combination on purpose
 * (see backend/app/schemas/submission.py).
 */
export function answerLocationWire(answer: LocalAnswer): AnswerLocationWire {
  if (answer.latitude == null || answer.longitude == null) {
    return {};
  }
  return {
    latitude: answer.latitude,
    longitude: answer.longitude,
    // 'gps_live' for a real device fix, 'user_selected' when this came from
    // the guide's currently selected TrailMind Location instead (see
    // AnswerQuestionScreen's placeToCapturedLocation). Falls back to
    // 'gps_live' for any answer synced before locationSource existed on
    // LocalAnswer -- every one of those WAS a live GPS fix in practice,
    // since LocationCaptureField was the only source of a captured answer
    // location at the time.
    location_source: answer.locationSource ?? 'gps_live',
    ...(answer.locationAccuracyMeters != null
      ? { location_accuracy_meters: answer.locationAccuracyMeters }
      : {}),
    ...(answer.locationCapturedAt ? { location_captured_at: answer.locationCapturedAt } : {}),
    ...(answer.locationLabel ? { location_label: answer.locationLabel } : {}),
    ...(answer.externalPlaceId ? { external_place_id: answer.externalPlaceId } : {}),
  };
}

File: backend/app/schemas/answer_location.py
"""Where the guide actually was when they answered a question.

WHY THIS EXISTS: an answer's Submission has always had its coordinates
derived server-side -- from the knowledge gap's own target coordinates
(app/services/question_answers.py) or from the place a place question is
about (app/services/place_question_answers.py). Neither is where the guide
stood, and when a guide has moved on, extraction's own fallback (their most
recent GuideLocation ping) is staler still: a guide who recorded a position
in a park and walked two kilometres before answering would have the answer
filed against the park.

So the client may now say where it actually was. Optional throughout: an
answer that carries no coordinates keeps exactly the pre-existing server-side
derivation, which is what every older app build sends and what a guide who
declines the location prompt still gets.

Validation deliberately mirrors SubmissionCreate's (app/schemas/submission.py)
rather than inventing a second dialect -- same tolerant unknown-source
degrading, same "coordinates are the evidence for a source label" rule.
"""

import logging
from datetime import datetime

from pydantic import BaseModel, Field, field_validator, model_validator

from app.db.models.submission import DEFAULT_LOCATION_SOURCE, SUBMISSION_LOCATION_SOURCES
from app.services.geo_validation import validate_optional_coordinate_pair

logger = logging.getLogger(__name__)


class AnswerLocationFields(BaseModel):
    """Mixed into both answer-create schemas. Every field optional."""

    latitude: float | None = Field(default=None, ge=-90, le=90)
    longitude: float | None = Field(default=None, ge=-180, le=180)
    location_source: str | None = None
    location_accuracy_meters: float | None = Field(default=None, gt=0)
    location_captured_at: datetime | None = None
    location_label: str | None = Field(default=None, max_length=255)
    external_place_id: str | None = Field(default=None, max_length=255)

    @field_validator("location_source")
    @classmethod
    def known_location_source(cls, value: str | None) -> str | None:
        # Tolerant, exactly like SubmissionCreate: an unrecognised label from
        # a newer/older app degrades to "unknown" rather than rejecting a
        # real answer over a provenance string.
        if value is None:
            return None
        cleaned = value.strip().lower()
        if cleaned not in SUBMISSION_LOCATION_SOURCES:
            logger.info(
                "Unknown location_source %r on an answer -- using %r.",
                value,
                DEFAULT_LOCATION_SOURCE,
            )
            return DEFAULT_LOCATION_SOURCE
        return cleaned

    @model_validator(mode="after")
    def validate_coordinate_pair(self) -> "AnswerLocationFields":
        if (self.latitude is None) != (self.longitude is None):
            raise ValueError("latitude and longitude must be supplied together")
        # Same (0, 0)-is-a-placeholder rule as SubmissionCreate -- see
        # app/services/geo_validation.py. A captured location that decodes to
        # exactly (0, 0) is rejected here rather than silently accepted as
        # 'gps_live'/'photo_exif' evidence it never actually was.
        validate_optional_coordinate_pair(self.latitude, self.longitude)
        if (
            self.location_source is not None
            and self.location_source != DEFAULT_LOCATION_SOURCE
            and self.latitude is None
        ):
            raise ValueError(
                f"location_source {self.location_source!r} requires latitude/longitude"
            )
        return self

    def has_captured_location(self) -> bool:
        """True only when a real coordinate pair arrived. `location_label` or
        `external_place_id` alone is not enough -- the coordinate is the
        evidence, and without it there is nothing to place the answer at."""
        return self.latitude is not None and self.longitude is not None

    def submission_location_kwargs(self) -> dict:
        """The Submission fields to use for a captured location. Callers use
        this ONLY when has_captured_location() is true, keeping their existing
        derivation untouched otherwise."""
        return {
            "latitude": self.latitude,
            "longitude": self.longitude,
            "location_source": self.location_source or DEFAULT_LOCATION_SOURCE,
            "location_accuracy_meters": self.location_accuracy_meters,
            "location_captured_at": self.location_captured_at,
            "location_label": self.location_label,
            "location_evidence": (
                "Captured by the guide on their device at the moment they answered."
            ),
            "external_place_id": self.external_place_id,
        }
